### ISSUE TITLE
Consul Server Crashing on DNS lookup

### ISSUE THREAD

**Issue by rmb938** [hashicorp/consul#2795]
### `consul version` for both Client and Server

```
Consul v0.7.5
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
```

### `consul info` for both Client and Server

Server
```
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 3
build:
        prerelease = 
        revision = '21f2d5a
        version = 0.7.5
consul:
        bootstrap = false
        known_datacenters = 1
        leader = false
        leader_addr = 192.168.23.101:8300
        server = true
raft:
        applied_index = 511360
        commit_index = 511360
        fsm_pending = 0
        last_contact = 65.233681ms
        last_log_index = 511360
        last_log_term = 119
        last_snapshot_index = 507928
        last_snapshot_term = 106
        latest_configuration = [{Suffrage:Voter ID:192.168.23.100:8300 Address:192.168.23.100:8300} {Suffrage:Voter ID:192.168.23.102:8300 Address:192.168.23.102:8300} {Suffrage:Voter ID:192.168.23.101:8300 Address:192.168.23.101:8300}]
        latest_configuration_index = 1
        num_peers = 2
        protocol_version = 1
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 119
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 98
        max_procs = 2
        os = linux
        version = go1.7.5
serf_lan:
        encrypted = true
        event_queue = 0
        event_time = 30
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 479
        members = 22
        query_queue = 0
        query_time = 1
serf_wan:
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1
        members = 1
        query_queue = 0
        query_time = 1
```

Client
```
agent:
        check_monitors = 0
        check_ttls = 1
        checks = 1
        services = 2
build:
        prerelease = 
        revision = '21f2d5a
        version = 0.7.5
consul:
        known_servers = 3
        server = false
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 43
        max_procs = 2
        os = linux
        version = go1.7.5
serf_lan:
        encrypted = true
        event_queue = 0
        event_time = 30
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 479
        members = 22
        query_queue = 0
        query_time = 1
```

### Operating system and Environment details

Centos 7 with 3 Consul Servers and ~20 consul clients

### Description of the Issue (and unexpected/desired result)

I added Vault to the Consul cluster configured to register it's own service.

When I lookup the service via DNS the Consul Server will randomly crash with the huge stack trace linked bellow and well as giving the following error when using Dig

```
[root@consul-server1 ~]# dig @127.0.0.1 -p 8600 vault.service.consul.rmb938.me 

; <<>> DiG 9.9.4-RedHat-9.9.4-38.el7_3.2 <<>> @127.0.0.1 -p 8600 vault.service.consul.rmb938.me
; (1 server found)
;; global options: +cmd
;; connection timed out; no servers could be reached
```

Querying any other service DNS record works fine without any issues. This only ever seems to happen when trying to lookup vault.

### Log Fragments or Link to [gist](https://gist.github.com/)

https://gist.github.com/rmb938/33f836c69c35a4373926cc386325dd3f

**Comment by slackpad** [comment]
Hi @rmb938 this looks like a recursor loop. Can you share some details about your Consul server configuration as well as what it shows for `/v1/catalog/service/vault?pretty`

**Comment by rmb938** [comment]
Well I don't event have dns recursing turned on so I am not really sure how that could be an issue.

Here is the consul server configuration:

```
{
  "log_level": "TRACE",
  "enable_syslog": true,
  "data_dir": "/var/lib/consul",
  "datacenter": "dc1",
  "domain": "consul.rmb938.me.",
  "encrypt": "<< REDACTED >>",
  "disable_remote_exec": true,
  "bind_addr": "192.168.23.100",
  "retry_join": [
    "192.168.23.101",
    "192.168.23.102",
    "192.168.23.100"
  ],
  "bootstrap_expect": 3,
  "ui": true,
  "server": true,
  "client_addr": "0.0.0.0",
  "performance": {
    "raft_multiplier": 1
  }
}
```

Here is the Consul Clients configuration:

```
{
  "enable_syslog": true,
  "data_dir": "/var/lib/consul",
  "datacenter": "dc1",
  "domain": "consul.rmb938.me.",
  "encrypt": "<< REDACTED  >>",
  "disable_remote_exec": true,
  "bind_addr": "192.168.23.112",
  "retry_join": [
    "192.168.23.101",
    "192.168.23.102",
    "192.168.23.100"
  ],
  "rejoin_after_leave": true
```

All agent's also have various service json configuration

```
{
  "service": {
    "name": "node_metrics",
    "tags": [
      "metrics"
    ],
    "port": 9100,
    "checks": [

    ]
  }
}
```

```
{
  "service": {
    "name": "consul_metrics",
    "tags": [
      "metrics"
    ],
    "port": 9107,
    "checks": [

    ]
  }
}
```

ect...

They also all have the following systemd unit file:

```
# /usr/lib/systemd/system/consul.service
[Unit]
Description=Consul Server Service

[Service]
TimeoutStartSec=0
Restart=always
ExecStart=/usr/bin/consul agent -config-dir=/etc/consul.d -data-dir=/var/lib/consul

[Install]
WantedBy=multi-user.target
```

**Comment by rmb938** [comment]
Thinking about it... this could be happening because the vault service address is set to `vault.service.consul.rmb938.me` causing infinite DNS lookups in consul. 

Per the Vault documentation it's "redirect_addr" should be set to the load balancer and if I am using consul dns for load balancing it would make sense for it to be set to the dns record. At least that is my thinking.

**Comment by slackpad** [comment]
Consul is trying to resolve that internal address and getting into a loop. We should guard against that in Consul and I'll /cc @sean- on here for thoughts on the right way forward from the Vault side.

**Comment by magiconair** [comment]
I can have a look. I was able to reproduce this on master as follows:

```
# start consul
consul agent -server -dev

# register service
curl -XPOST -d '{"ID":"foo1","Name": "foo","Address":"foo.service.consul","Port": 8000,"Check": {"Script": "/usr/bin/true", "Interval":"10s"}}' localhost:8500/v1/agent/service/register

# add an entry to /etc/hosts (without that consul won't crash)
127.0.0.1 foo.service.consul

# lookup and boom!
dig @127.0.0.1 -p 8600 foo.service.consul.
```

**Comment by magiconair** [comment]
After thinking about this for a bit I think the underlying issue is `vault` (or any service that can already be discovered via consul DNS) using its consul DNS name as the `ServiceAddress`. This is a loop by definition. This makes sense for an external service IMO but not for an internal one. 

But maybe I'm missing something here. 

`consul` still shouldn't crash though.

**Comment by slackpad** [comment]
@magiconair you are right - this is a misconfiguration. I think your fix in #3082 is about as simple as we can get to help prevent loops on the Consul side for cases like this, and is worth adding. I don't think Vault can really detect this because it doesn't know the Consul suffix, and it would be weird to enforce this at registration time in Consul, though that's a possibility. Opened https://github.com/hashicorp/vault/issues/2783 over there to make sure we have a clear story on how to configure Vault w/Consul in the Vault documentation.
